Short-Answer/Algorithms_Answers.md

```diff
@@ -1 +1,55 @@
-Add your answers to the Algorithms exercises here.
\ No newline at end of file
+Add your answers to the Algorithms exercises here.
+
+# Analysis of Algorithms
+
+## Exercise I
+
+Give an analysis of the running time of each snippet of
+pseudocode with respect to the input size n of each of the following:
+
+```
+a)  a = 0 # So we start at 0...
+    while (a < n * n * n): # ...then establish a condition of a cube...
+      a = a + n * n # and try to meet it with a square.
+```
+If n is 0 then we're done, but if n is 1, then we loop once, and if n is 2, then we loop twice, and then if n is 3, we loop 3 times. If n is 10, we loop 10 times, etc. This is an O(n) algorithm, in spite of its confusing expression.
+
+
+
+```
+b)  sum = 0
+    for i in range(n):
+      i += 1 # aaa...
+      for j in range(i + 1, n): # aaaaaa
+        j += 1
+        for k in range(j + 1, n): #aaaaaaAAAAAAA
+          k += 1
+          for l in range(k + 1, 10 + k): #AAAAAAああああああああああああああああ
+            l += 1
+            sum += 1
+```
+This algorithm is (internal screaming).
+
+The outer "i" loop finishes in half the time of n, because it increments twice. But each iteration of i brings with it an iteration of j, which is also half the time. Then we iterate through k, which will also be half. And we bring with us an iteration of l, reliably about 5 times (because if k is 1, we go from 2 to 11, but if k is 1000, we go from 1001 to 1010). Could be 4, but that doesn't seem terribly important at this point, it's sufficient to know that innermost l loop does not actually increase with the size of n.
+
+So if we fire this up with an n of 10 or so (I'm barely doing the math at this point, yes, because I don't care about the exact values now that I've seen the pattern), we have about 5 iterations of i (scales directly in relation to n), each of which brings about 5, then 4, then 3, then 2, then 1 iterations of j, as j loops over a smaller amount each time (an n log^n scale), each of which does something similar with k (another n log^n scale), and then brings another 4 or 5 iterations of l, just as a gotcha.
+
+
+This is something nightmarish. It explodes. I am not sure it is n^factorial, but it is O 2^n at least, and it's at least trying to make n^factorial feel nervous. I'll get a more exact number if I finish robot sort early.
+
```

Comment:
Simplify the thought process down a bit for this one.
We have 4 nested for-loops.  3 of which have a number of iterations dependent on n. 
Therefore we approximate it as O(n^3).

Even though the second and third loops start from i and j for each iteration, we still consider this O(n) for the individual loop even though it doesn't end up going through n total iterations each loop.